
---
title: "PathwayGuidedRF"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include = FALSE, eval = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

This R package provides functions for the identification of important pathways
or gene sets using multiple pathway guided random forest (RF) approaches. 
Furthermore, it includes functions to simulate pathway based gene expression
data under two different scenarios. 

<!-- Please cite the following manuscript if you use the package:   -->
<!-- S Seifert, S Gundlach, O Junge and S Szymczak (2019) Integrating biological  -->
<!-- knowledge and omics data using pathway guided random forests: a benchmarking  -->
<!-- study. Submitted to Genome Biology. -->

## Installation

<!-- install.packages(c("ranger", "Boruta", "Umpire", "geoR", "MASS")) -->

You can install the most recent version of PathwayGuidedRF from GitHub as 
follows:

```{r, eval = FALSE}
library(devtools)
install_github("szymczak-lab/PathwayGuidedRF")
```

## Simulated data

We will use an example data set simulated under the scenario of simulation 
study 1 as described in the paper. 

First, we load the package:

```{r}
library(PathwayGuidedRF)
```

We then specify some characteristics of the three pathways we would like to 
simulate.
Note that the first two pathways contain some differentially expressed genes,
while the third pathway has no signal. A well performing pathway analysis 
approach should thus identify the first two pathways.

```{r}
# define pathway parameters
info.par = data.frame(pw = paste0("pw", 1:3),
                      no.genes = c(100, rep(20, 2)),
                      prop.de = c(1, 0.5, 0),
                      cor = rep(0.6, 3),
                      stringsAsFactors = FALSE)
```

We finally simulate 30 individuals, of which 15 should be cases. In order to be 
able to reproduce the data and the results we set the random number generator to 
the specified seed. The parameter gamma.range defines the the range of absolute
effect sizes for the differential expression.

```{r, message = FALSE}
set.seed(12345)
sim.data.l = sim.data.study.1(info.par = info.par,
                              no.samples = 30,
                              no.cases = 15,
                              seed = 42,
                              gamma.range = c(0.5, 1.5))
```

The resulting list sim.data.l contains a training and a test data set simulated
under the same setting (i.e. same genes and effect sizes). We can check the
dimensions (30 individuals and 140 genes). Note that the first column contains 
the outcome coded as 0 (= control) and 1 (= case).

```{r}
data.train = sim.data.l$data.train
dim(data.train)
head(data.train[, 1:5])
table(data.train$y)
```

## Pathway analysis using RF

We can perform RF based pathway analysis using several approaches that are 
implemented in the package. The results have a similar structure but also 
contain some method specific information.

```{r}
# synthetic feature approach
set.seed(12345)
res.sf = pw.rf.synthetic.features(x = sim.data.l$data.train[, -1],
                                  y = sim.data.l$data.train[, 1],
                                  info.pw = sim.data.l$info.pw,
                                  type = "classification",
                                  not.assoc.pw = FALSE)

# selected pathways
res.sf$pw.sel

# more information for each pathway
res.sf$results.pw

# hunting approach
set.seed(12345)
res.hunt = pw.rf.hunt(x = sim.data.l$data.train[, -1],
                      y = sim.data.l$data.train[, 1],
                      info.pw = sim.data.l$info.pw,
                      type = "classification",
                      not.assoc.pw = FALSE)
res.hunt$pw.sel
res.hunt$results.pw

```

For illustration we use the default of 20 permutations for the 
prediction error method. However, since we analyse only three pathways the total 
number of permutations is 60 which is too small for reliable results. In a real 
analysis the total number of permutation should be larger than 1000 which is 
often already achieved with 20 permutations per pathway since many more pathways
are usually tested.

```{r}
# prediction error approach
set.seed(12345)
res.pe = pw.rf.pred.error(x = sim.data.l$data.train[, -1],
                          y = sim.data.l$data.train[, 1],
                          info.pw = sim.data.l$info.pw,
                          type = "classification",
                          not.assoc.pw = FALSE,
                          no.perm = 20)
res.pe$pw.sel
res.pe$results.pw

```

A warning is issued for the LeFE approach since not enough genes outside of the
first pathway are available in our small example data set. Thus, no P value
could be calculated for this pathway.

```{r}
# LeFE approach
set.seed(12345)
res.lefe = pw.rf.lefe(x = sim.data.l$data.train[, -1],
                      y = sim.data.l$data.train[, 1],
                      info.pw = sim.data.l$info.pw,
                      type = "classification",
                      not.assoc.pw = FALSE)
res.lefe$pw.sel
res.lefe$results.pw

```
